cliquet/storage/postgresql/__init__.py

```diff
@@ -0,0 +1,458 @@
+import contextlib
+import json
+import os
+from collections import defaultdict
+
+import psycopg2
+import psycopg2.extras
+import six
+from six.moves.urllib import parse as urlparse
+
+from cliquet import logger
+from cliquet import errors
+from cliquet.storage import StorageBase, exceptions
+from cliquet.utils import classname, COMPARISON
+
+
+psycopg2.extensions.register_type(psycopg2.extensions.UNICODE)
+psycopg2.extensions.register_type(psycopg2.extensions.UNICODEARRAY)
+
+
+DEFAULT_MAX_FETCH_SIZE = 10000
+
+
+class PostgreSQL(StorageBase):
+    """Storage backend for PostgreSQL.
+    """
+    def __init__(self, *args, **kwargs):
+        super(PostgreSQL, self).__init__(*args, **kwargs)
+        self._max_fetch_size = kwargs.pop('max_fetch_size')
+        self._conn_kwargs = kwargs
+        self._init_schema()
+
+    @contextlib.contextmanager
+    def connect(self):
+        """Connect to the database and instantiates a cursor.
+        At exiting the context manager, a COMMIT is performed on the current
+        transaction if everything went well. Otherwise transaction is ROLLBACK,
+        and everything cleaned up.
+
+        If the database could not be be reached a 503 error is raised.
+        """
+        conn = None
+        cursor = None
+        try:
+            conn = psycopg2.connect(**self._conn_kwargs)
+            options = dict(cursor_factory=psycopg2.extras.DictCursor)
+            cursor = conn.cursor(**options)
+            yield cursor
+            conn.commit()
+        except psycopg2.Error as e:
+            if cursor:
+                logger.debug(cursor.query)
+            logger.exception(e)
+            if conn and not conn.closed:
+                conn.rollback()
+            if isinstance(e, psycopg2.OperationalError):
+                raise errors.HTTPServiceUnavailable()
+            raise
+        finally:
+            if cursor:
+                cursor.close()
+            if conn and not conn.closed:
+                conn.close()
+
+    def _init_schema(self):
+        """Create PostgreSQL tables, only if not exists.
+
+        :note:
+            Relies on JSON fields, available in recent versions of PostgreSQL.
+        """
+        # Since indices cannot be created with IF NOT EXISTS, inspect:
+        try:
+            inspect_tables = "SELECT * FROM records LIMIT 0;"
```

Comment:
I expected postgresql to come with a way to know if a table existed, by looking this up in a meta table somewhere, isn't it possible?